### PR TITLE
Az/custom annotation

### PR DESCRIPTION
**Pull request by azhavoro**

### CODE REVIEW

File: cvat/apps/custom_annotation/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+Application uses OpenCV dnn module with DLDT backed for inference.
+
+### Usage
+To annotate task with custom model you need prepare 4 files:
+1. **Model config** - a text file contains network configuration. It could be a file with the following extensions:
+   * *.xml (DLDT)
+1. **Model weights** - a binary file contains trained weights. The following file extensions are expected for models from different frameworks:
+   * *.bin (DLDT)
+1. **Preprocessing configureation and label map** - simple json file that describes image dimentions and preprocessing options. For more details please view [OpenCV](https://docs.opencv.org/3.4/d6/d0f/group__dnn.html#ga0b7b7c3c530b747ef738178835e1e70f) documentation.  
+Label values in label_map should be exactly equal to labels wich task was created, otherwise will be ignored.
```
Comment (nmanovic) [inline cvat/apps/custom_annotation/README.md]: some typos in the sentence. Please check with an editor.

File: cvat/apps/custom_annotation/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
+## Custom annotation
```
Comment (nmanovic) [inline cvat/apps/custom_annotation/README.md]: Let's rename the app custom_annotation -> auto_annotator or just annotator

File: cvat/apps/custom_annotation/views.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    model =  ModelLoader(path_to_model=model, blob_params=blob_params)
+    model.load()
+
+    return model, class_names
+
+def process_detections(detections, path_to_conv_script):
+    import importlib.util
+    spec = importlib.util.spec_from_file_location('converter', path_to_conv_script)
+    converter = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(converter)
```
Comment (nmanovic) [inline cvat/apps/custom_annotation/views.py]: Is it save to run the line? Do we need a special "sandbox" solution to run custom code? (https://stackoverflow.com/questions/3068139/how-can-i-sandbox-python-in-pure-python)

Comment (nmanovic) [reply cvat/apps/custom_annotation/views.py]: Just look at available options. In trivial case we can use `exec somePythonCode in {'__builtins__': {}}, {}` together with 'admin' right to add a new model.

File: cvat/apps/auto_annotation/README.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+This application will be enabled automatically if OpenVINO component was installed and allows to use custom detection models for preannotation.
+Supported frameworks:
+* DLDT form OpenVINO toolkit
+
+Application uses OpenCV dnn module with DLDT backed for inference.
+
+### Usage
+To annotate task with custom model you need prepare 4 files:
+1. **Model config** - a text file that contains network configuration. It could be a file with the following extension:
```
Comment (bsekachev) [inline cvat/apps/auto_annotation/README.md]: could or should? Does OpenVino support other formats?

File: cvat/apps/auto_annotation/README.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+### Description
+
+This application will be enabled automatically if OpenVINO component was installed and allows to use custom detection models for preannotation.
+Supported frameworks:
+* DLDT form OpenVINO toolkit
+
+Application uses OpenCV dnn module with DLDT backed for inference.
+
+### Usage
+To annotate task with custom model you need prepare 4 files:
```
Comment (bsekachev) [inline cvat/apps/auto_annotation/README.md]: a task with a custom model..

File: cvat/apps/auto_annotation/model_loader.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+import cv2
```
Comment (bsekachev) [inline cvat/apps/auto_annotation/model_loader.py]: # Copyright (C) 2018 Intel Corporation
#
# SPDX-License-Identifier: MIT

File: cvat/apps/auto_annotation/views.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        write_file(config_file_path, config_file)
+
+        convertation_file = request.FILES['conv_script']
+        convertation_file_path = os.path.join(upload_dir, convertation_file.name)
+        write_file(convertation_file_path, convertation_file)
+
+
+        db_labels = db_task.label_set.prefetch_related('attributespec_set').all()
+        db_labels = {db_label.id:db_label.name for db_label in db_labels}
+
```
Comment (bsekachev) [inline cvat/apps/auto_annotation/views.py]: I think we need to confirm an user immediately:
I know, this check will later, but if queue is busy, user can waiting long for bad response. 
```
if not db_labels:
    return HttpResponseBadRequest('No labels found for an auto annotation')
```

Comment (bsekachev) [reply cvat/apps/auto_annotation/views.py]: UPD: in this case we need move an labels intersection search here

File: cvat/apps/auto_annotation/README.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
+## Auto annotation
+
+### Description
+
+This application will be enabled automatically if OpenVINO&trade; component is installed. It allows to use custom models for preannotation.
+Supported only DLDT framework from OpenVINO&trade; toolkit. If you woold like to annotate task with custom model please convert it to the
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: would (please use a spell checker for the whole document)

File: cvat/apps/auto_annotation/README.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+intermediate representation (IR) format via model optimizer tool.
+See [OpenVINO documentation](https://software.intel.com/en-us/articles/OpenVINO-InferEngine) for details.
+
+### Usage
+To annotate task with a custom model you need prepare 4 files:
+1. __Model config__ - a text file that contains network configuration. The following file extension is expected:
+   * __*.xml__
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: Please remove `* __*.xml__`. Instead just write `__Model config__ (*.xml) - ...`

File: cvat/apps/auto_annotation/README.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+1. __Model weights__ - a binary file that contains trained weights. The following file extension is expected:
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: `__Model weights__ (.bin) - ...`

File: cvat/apps/auto_annotation/README.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      "polygon_paths": [],
+      "polyline_paths": [],
+      "points_paths": [],
+    }
+    ```
+
+    Example for SSD based network:
+    ```python
+    def process_results(detections, results):
+    def clip(value):
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: I believe the line has incorrect indentation

File: cvat/apps/auto_annotation/README.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        "15": "person",
+        "16": "pottedplant",
+        "17": "sheep",
+        "18": "sofa",
+        "19": "train",
+        "20": "tvmonitor"
+      }
+    }
+    ```
+1. __Interpretation script__ - python file that used to convert output results from net to CVAT format. This code running inside restricted environment.
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: `__Interpretation script__ (*.py) - ...`

File: cvat/apps/auto_annotation/README.md [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+intermediate representation (IR) format via model optimizer tool.
+See [OpenVINO documentation](https://software.intel.com/en-us/articles/OpenVINO-InferEngine) for details.
+
+### Usage
+To annotate task with a custom model you need prepare 4 files:
+1. __Model config__ - a text file that contains network configuration. The following file extension is expected:
+   * __*.xml__
+1. __Model weights__ - a binary file that contains trained weights. The following file extension is expected:
+   * __*.bin__
+1. __Label map__ - simple json file that contains the `label_map` dictionary like object with string values for label numbers.
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: `__Label map__ (*.json) - ...`

File: cvat/apps/auto_annotation/README.md [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+    Example for SSD based network:
+    ```python
+    def process_results(detections, results):
+    def clip(value):
+        return max(min(1.0, value), 0.0)
+
+    boxes = results['boxes']
+
+    for frame_results in detections:
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: How the function will look for the model: file:///opt/intel/computer_vision_sdk_2018.5.445/deployment_tools/intel_models/person-detection-retail-0002/description/person-detection-retail-0002.html?

File: cvat/apps/auto_annotation/README.md [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        "16": "pottedplant",
+        "17": "sheep",
+        "18": "sofa",
+        "19": "train",
+        "20": "tvmonitor"
+      }
+    }
+    ```
+1. __Interpretation script__ - python file that used to convert output results from net to CVAT format. This code running inside restricted environment.
+List of builtins functions that available to use:
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: to use: __str, int, float, max, min, range, ...__

File: cvat/apps/auto_annotation/README.md [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+  frame_width = frame_results["frame_width"]
+  frame_number = frame_results["frame_id"]
+
+  for i in range(frame_results["detections"].shape[2]):
+    confidence = frame_results["detections"][0, 0, i, 2]
+    if confidence < 0.5:
+      continue
+
+    results.add_box(
+      xtl="{:.2f}".format(clip(frame_results["detections"][0, 0, i, 3]) * frame_width),
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: expected values are float or int

File: cvat/apps/auto_annotation/README.md [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+      attributes={
+        "confidence": "{:.2f}".format(confidence),
+      },
+    )
+```
+
+
+#### [Landmarks-regression-retail-0009](https://github.com/opencv/open_model_zoo/blob/2018/intel_models/landmarks-regression-retail-0009/description/landmarks-regression-retail-0009.md) (OpenVINO toolkit)
+
+__Note__: Model weights are available in OpenVINO redistributable package.
```
Comment (nmanovic) [inline cvat/apps/auto_annotation/README.md]: Model configuration (*.xml) and weights (*.bin) ...

**Comment by nmanovic** [comment]
@azhavoro What is wrong? Also no problems are reported in UI.

2018-12-26 15:29:41,834 DEBG 'rqworker_low' stderr output:
/home/django/cvat/apps/auto_annotation/model_loader.py:30: DeprecationWarning: from_ir() method of IENetwork is deprecated. Please use IENetwork class constructor to create valid IENetwork instance
  network = IENetwork.from_ir(model=self._model, weights=self._weights)
[2018-12-26 15:29:41,834] ERROR cvat.server.task_6: exception was occurred during auto annotation of the task
Traceback (most recent call last):
  File "/home/django/cvat/apps/auto_annotation/views.py", line 242, in create_thread
    update_progress=update_progress,
  File "/home/django/cvat/apps/auto_annotation/views.py", line 199, in run_inference_engine_annotation
    "detections": model.infer(frame),
  File "/home/django/cvat/apps/auto_annotation/model_loader.py", line 44, in infer
    _, _, h, w = self._input_layout
TypeError: 'openvino.inference_engine.ie_api.InputInfo' object is not iterable

**Comment by azhavoro** [comment]
@nmanovic IE python API changes are cause of issue. I've added fix and it works with R3 and R5. I suggest specify supported version of OpenVINO in documentation at least.

**Pull request merged**